### PR TITLE
Collects failed assets on LyTestTools test failures

### PR DESCRIPTION
**Pull request by evanchia-ly-sdets**
Tested by running locally with failed assets. Unit tests pass.

Signed-off-by: evanchia <evanchia@amazon.com>

### CODE REVIEW

**Review by AMZN-scspaldi — APPROVED** [review]

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py [hunk 1]
```diff
@@ -171,23 +171,22 @@ def save_failed_asset_joblogs(workspace: AbstractWorkspace) -> None:
 
 def _check_log_errors_warnings(log_path: str) -> bool:
     """
-    Checks to see if the asset log contains any errors or warnings by reading the last line of the file.
-    Example log line: '~~1643759303647~~1~~00000000000009E0~~AssetBuilder~~S: 0 errors, 1 warnings'
+    Checks to see if the asset log contains any errors or warnings by reading the last line of the file. There are
+    multiple types of log regex. Returns True is no regex is found because something probably went wrong.
+    Example log lines: ~~1643759303647~~1~~00000000000009E0~~AssetBuilder~~S: 0 errors, 1 warnings
+    (Current issue where the expected line isn't always the last line. This function will instead query each line until
+    the issue is resolved).
```
Comment (Kadino) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py:178]: Please remove this stray comment or add a reference to an open GHI ticket

This appears to be docstring thrash. Would be better to not claim specifics about how the regex detects the errors in this docstring, and leave the code to track implementation details. In this case, you can also edit the return value to more simply state "True if an error or warning is detected, else False"

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py [hunk 2]
```diff
@@ -171,23 +171,22 @@ def save_failed_asset_joblogs(workspace: AbstractWorkspace) -> None:
...
 
     :param log_path: The pull path to the asset log file to read
```
Comment (Kadino) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py:180]: pull ==> full ?

**Review by Kadino — CHANGES_REQUESTED** [review]
Just to check: this uploads logs of assets, not the assets themselves?

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py (hunk 2 above)
Comment (evanchia-ly-sdets) [reply Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py]: fixed

**Comment by evanchia-ly-sdets** [comment]
> Just to check: this is for collecting logs of assets, not the assets themselves?

correct

**Review by Kadino — APPROVED** [review]

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py (hunk 1 above)
Comment (Kadino) [reply Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py]: I don't think a new ticket is necessary, unless this wider regex parsing is **spectacularly** slow.

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test.py [hunk 3]
```diff
@@ -961,6 +964,11 @@ def _run_batched_tests(self, request: Request, workspace: AbstractWorkspace, edi
                                               extra_cmdline_args)
         assert results is not None
         editor_test_data.results.update(results)
+        # If a single test did not pass, save assets with errors and warnings
```
Comment (Kadino) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test.py:967]: "a single" => "at least one"

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py [hunk 4]
```diff
@@ -151,3 +152,42 @@ def retrieve_last_run_test_index_from_output(test_spec_list: list[EditorTestBase
...
+
+def save_failed_asset_joblogs(workspace: AbstractWorkspace) -> None:
+    """
+    Checks all asset logs in the JobLogs directory to see if the asset has any warnings or errors. If so, the asset is
+    saved via ArtifactManager.
+
+    :param workspace: The AbstractWorkspace to access the JobLogs path
+    :return: None
+    """
+    # Tuples are in the form [this directory, [child directories], [files in this directory]]
```
Comment (Kadino) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py:164]: Not necessary to document os.walk here, unless something uncommon is being done and needs explanation

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py [hunk 5]
```diff
@@ -151,3 +152,42 @@ def retrieve_last_run_test_index_from_output(test_spec_list: list[EditorTestBase
...
+    for walk_tuple in os.walk(workspace.paths.ap_job_logs()):
+        for log_file in walk_tuple[2]:
+            full_log_path = os.path.join(walk_tuple[0], log_file)
+            # Only save asset logs that contain errors or warnings
+            if _check_log_errors_warnings(full_log_path):
+                try:
+                    workspace.artifact_manager.save_artifact(full_log_path)
+                except Exception as e:  # Purposefully broad
+                    logger.warning(e)
```
Comment (Kadino) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py:173]: This should log a human-readable warning that the artifact (at the given path) is being skipped, due to the error. Else the logging has little context.

I dislike the pattern of broad catch-and-discard, but this cleanup script is in fact an appropriate place for it.

**Review by Kadino — APPROVED** [review]
Feedback below are recommended changes which don't need to block moving forward with this PR.

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py (hunk 5 above)
Comment (evanchia-ly-sdets) [reply Tools/LyTestTools/ly_test_tools/o3de/editor_test_utils.py:173]: improved logging

**Pull request merged**